File: web/src/contracts/pageContext.ts
export const PAGE_CONTEXT_FIELD_NAMES = [
  'surface_id',
  'organization_scope',
  'work_object_refs',
  'source_refs',
  'filters',
  'selected_metric',
  'allowed_capabilities',
  'freshness',
  'visibility',
] as const;

const MAX_REFERENCES = 200;
const MAX_FILTERS = 32;
const MAX_CAPABILITIES = 64;
export const PAGE_CONTEXT_TIMESTAMP_MAX_LENGTH = 64;
export const PAGE_CONTEXT_TIMESTAMP_PATTERN = String.raw`^\d{4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[12]\d|3[01])T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d{1,6})?Z$`;
const surfaceIdPattern = /^[a-z0-9]+(?:[._-][a-z0-9]+)*$/;
const identifierPattern = /^[A-Za-z0-9]+(?:[._:@/-][A-Za-z0-9]+)*$/;
const capabilityIdPattern = /^[a-z0-9]+(?:[._:-][a-z0-9]+)*$/;
const domMarkupPattern =
  /<!doctype\s+html|<\s*\/?\s*(?:html|body|form|input|script|table|div|main)\b/i;
export const PAGE_CONTEXT_SENSITIVE_VALUE_PATTERNS = [
  String.raw`(?<!\d)(?:\d{17}[\dXx]|\d{15})(?!\d)`,
  String.raw`bearer\s+\S+`,
  String.raw`\b[a-z][a-z0-9+.-]*://[^\s/@]+@`,
  String.raw`(?:authorization|session(?:[\s_-]?id)?|access[\s_-]?token|refresh[\s_-]?token|set[\s_-]?cookie|cookie|password|passwd|api[\s_-]?key|secret|client[\s_-]?secret|private[\s_-]?key|loginid|userpassword|oa[\s_-]?userid|userid)\s*[:=]\s*\S+`,
] as const;
const sensitiveValuePatterns = PAGE_CONTEXT_SENSITIVE_VALUE_PATTERNS.map(
  (pattern) => new RegExp(pattern, 'i'),
);
const timestampPattern = new RegExp(PAGE_CONTEXT_TIMESTAMP_PATTERN);

type PageVisibility = 'principal' | 'department' | 'organization';
type PageFilterOperator =
  | 'equals'
  | 'not_equals'
  | 'contains'
  | 'greater_than'
  | 'less_than';
type PageFilterValue = string | number | boolean;
type PageFreshnessState = 'unknown' | 'reported' | 'stale';

export interface OrganizationScope {
  readonly tenant_id: string;
  readonly organization_id: string | null;
  readonly department_id: string | null;
}

export interface WorkObjectReference {
  readonly work_object_id: string;
}

export interface SourceReference {
  readonly source_system: string;
  readonly source_ref: string;
}

export interface PageFilter {
  readonly field: string;
  readonly operator: PageFilterOperator;
  readonly value: PageFilterValue;
  readonly source: 'visible_control';
}

export interface PageFreshness {
  readonly state: PageFreshnessState;
  readonly observed_at: string | null;
}

/** Page-declared data only. It is never a backend authorization result. */
export interface PageContextDeclaration {
  readonly surface_id: string;
  readonly organization_scope: OrganizationScope | null;
  readonly work_object_refs: readonly WorkObjectReference[];
  readonly source_refs: readonly SourceReference[];
  readonly filters: readonly PageFilter[];
  readonly selected_metric: string | null;
  readonly allowed_capabilities: readonly string[];
  readonly freshness: PageFreshness;
  readonly visibility: PageVisibility;
}

export class PageContextValidationError extends Error {
  readonly code = 'invalid_page_context';

  constructor(path: string, reason: string) {
    super(`${path}: ${reason}`);
    this.name = 'PageContextValidationError';
  }
}

function fail(path: string, reason: string): never {
  throw new PageContextValidationError(path, reason);
}

function asRecord(value: unknown, path: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    return fail(path, 'must be an object');
  }
  return value as Record<string, unknown>;
}

function assertExactKeys(
  value: Record<string, unknown>,
  expected: readonly string[],
  path: string,
) {
  const actual = Object.keys(value);
  const missing = expected.filter(
    (key) => !Object.prototype.hasOwnProperty.call(value, key),
  );
  const extra = actual.filter((key) => !expected.includes(key));
  if (missing.length > 0 || extra.length > 0) {
    fail(path, 'has missing or unrecognized fields');
  }
}

function safeText(value: unknown, path: string, maxLength: number): string {
  if (typeof value !== 'string' || value.length === 0 || value.length > maxLength) {
    return fail(path, `must be a non-empty string of at most ${maxLength} characters`);
  }
  if (
    [...value].some((character) => {
      const codePoint = character.codePointAt(0) ?? 0;
      return codePoint <= 31 || codePoint === 127;
    })
  ) {
    return fail(path, 'must contain only printable single-line text');
  }
  if (domMarkupPattern.test(value)) {
    return fail(path, 'must not contain DOM markup');
  }
  if (sensitiveValuePatterns.some((pattern) => pattern.test(value))) {
    return fail(path, 'must not contain credential-like material');
  }
  return value;
}

function identifier(value: unknown, path: string): string {
  const parsed = safeText(value, path, 256);
  if (!identifierPattern.test(parsed)) {
    return fail(path, 'contains unsupported identifier characters');
  }
  return parsed;
}

function surfaceId(value: unknown): string {
  const parsed = safeText(value, 'surface_id', 80);
  if (!surfaceIdPattern.test(parsed)) {
    return fail('surface_id', 'must use lowercase slug form');
  }
  return parsed;
}

function capabilityId(value: unknown, path: string): string {
  const parsed = safeText(value, path, 120);
  if (!capabilityIdPattern.test(parsed)) {
    return fail(path, 'must use lowercase dotted slug form');
  }
  return parsed;
}

function arrayValue(value: unknown, path: string, maxLength: number): unknown[] {
  if (!Array.isArray(value) || value.length > maxLength) {
    return fail(path, `must be an array with at most ${maxLength} items`);
  }
  return value;
}

function optionalIdentifier(value: unknown, path: string): string | null {
  return value === null ? null : identifier(value, path);
}

function organizationScope(value: unknown): OrganizationScope | null {
  if (value === null) {
    return null;
  }
  const raw = asRecord(value, 'organization_scope');
  assertExactKeys(
    raw,
    ['tenant_id', 'organization_id', 'department_id'],
    'organization_scope',
  );
  const parsed = {
    tenant_id: identifier(raw.tenant_id, 'organization_scope.tenant_id'),
    organization_id: optionalIdentifier(
      raw.organization_id,
      'organization_scope.organization_id',
    ),
    department_id: optionalIdentifier(
      raw.department_id,
      'organization_scope.department_id',
    ),
  };
  if (parsed.department_id !== null && parsed.organization_id === null) {
    return fail('organization_scope', 'department_id requires organization_id');
  }
  return Object.freeze(parsed);
}

function workObjectRefs(value: unknown): readonly WorkObjectReference[] {
  const parsed = arrayValue(value, 'work_object_refs', MAX_REFERENCES).map(
    (item, index) => {
      const raw = asRecord(item, `work_object_refs[${index}]`);
      assertExactKeys(raw, ['work_object_id'], `work_object_refs[${index}]`);
      return Object.freeze({
        work_object_id: identifier(
          raw.work_object_id,
          `work_object_refs[${index}].work_object_id`,
        ),
      });
    },
  );
  if (new Set(parsed.map((item) => item.work_object_id)).size !== parsed.length) {
    return fail('work_object_refs', 'must not contain duplicates');
  }
  return Object.freeze(parsed);
}

function sourceRefs(value: unknown): readonly SourceReference[] {
  const parsed = arrayValue(value, 'source_refs', MAX_REFERENCES).map(
    (item, index) => {
      const raw = asRecord(item, `source_refs[${index}]`);
      assertExactKeys(
        raw,
        ['source_system', 'source_ref'],
        `source_refs[${index}]`,
      );
      return Object.freeze({
        source_system: identifier(
          raw.source_system,
          `source_refs[${index}].source_system`,
        ),
        source_ref: safeText(
          raw.source_ref,
          `source_refs[${index}].source_ref`,
          256,
        ),
      });
    },
  );
  const keys = parsed.map((item) => `${item.source_system}\u0000${item.source_ref}`);
  if (new Set(keys).size !== keys.length) {
    return fail('source_refs', 'must not contain duplicates');
  }
  return Object.freeze(parsed);
}

function filters(value: unknown): readonly PageFilter[] {
  const operators: readonly PageFilterOperator[] = [
    'equals',
    'not_equals',
    'contains',
    'greater_than',
    'less_than',
  ];
  return Object.freeze(
    arrayValue(value, 'filters', MAX_FILTERS).map((item, index) => {
      const path = `filters[${index}]`;
      const raw = asRecord(item, path);
      assertExactKeys(raw, ['field', 'operator', 'value', 'source'], path);
      if (
        typeof raw.operator !== 'string' ||
        !operators.includes(raw.operator as PageFilterOperator)
      ) {
        return fail(`${path}.operator`, 'is not a supported operator');
      }
      if (raw.source !== 'visible_control') {
        return fail(`${path}.source`, 'must identify a visible control');
      }
      return Object.freeze({
        field: identifier(raw.field, `${path}.field`),
        operator: raw.operator as PageFilterOperator,
        value: filterValue(raw.value, `${path}.value`),
        source: 'visible_control' as const,
      });
    }),
  );
}

function filterValue(value: unknown, path: string): PageFilterValue {
  if (typeof value === 'string') {
    return safeText(value, path, 500);
  }
  if (typeof value === 'boolean') {
    return value;
  }
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }
  return fail(path, 'must be a finite scalar value');
}

function allowedCapabilities(value: unknown): readonly string[] {
  const parsed = arrayValue(
    value,
    'allowed_capabilities',
    MAX_CAPABILITIES,
  ).map((item, index) => capabilityId(item, `allowed_capabilities[${index}]`));
  if (new Set(parsed).size !== parsed.length) {
    return fail('allowed_capabilities', 'must not contain duplicates');
  }
  return Object.freeze(parsed);
}

function freshness(value: unknown): PageFreshness {
  const raw = asRecord(value, 'freshness');
  assertExactKeys(raw, ['state', 'observed_at'], 'freshness');
  const states: readonly PageFreshnessState[] = ['unknown', 'reported', 'stale'];
  if (typeof raw.state !== 'string' || !states.includes(raw.state as PageFreshnessState)) {
    return fail('freshness.state', 'is not a supported freshness state');
  }
  const observedAt =
    raw.observed_at === null
      ? null
      : safeText(
          raw.observed_at,
          'freshness.observed_at',
          PAGE_CONTEXT_TIMESTAMP_MAX_LENGTH,
        );
  if (raw.state !== 'unknown' && observedAt === null) {
    return fail('freshness', 'reported or stale state requires observed_at');
  }
  if (observedAt !== null && !isValidTimestamp(observedAt)) {
    return fail(
      'freshness.observed_at',
      'must be a valid UTC RFC3339 timestamp',
    );
  }
  return Object.freeze({
    state: raw.state as PageFreshnessState,
    observed_at: observedAt,
  });
}

function isValidTimestamp(value: string): boolean {
  if (!timestampPattern.test(value)) {
    return false;
  }
  const year = Number(value.slice(0, 4));
  const month = Number(value.slice(5, 7));
  const day = Number(value.slice(8, 10));
  const hour = Number(value.slice(11, 13));
  const minute = Number(value.slice(14, 16));
  const second = Number(value.slice(17, 19));
  if (year === 0) {
    return false;
  }
  const parsed = new Date(0);
  parsed.setUTCFullYear(year, month - 1, day);
  parsed.setUTCHours(hour, minute, second, 0);
  return (
    parsed.getUTCFullYear() === year &&
    parsed.getUTCMonth() === month - 1 &&
    parsed.getUTCDate() === day &&
    parsed.getUTCHours() === hour &&
    parsed.getUTCMinutes() === minute &&
    parsed.getUTCSeconds() === second
  );
}

function visibility(value: unknown): PageVisibility {
  if (value === 'principal' || value === 'department' || value === 'organization') {
    return value;
  }
  return fail('visibility', 'is not a supported visibility');
}

export function parsePageContext(value: unknown): PageContextDeclaration {
  const raw = asRecord(value, 'page_context');
  assertExactKeys(raw, PAGE_CONTEXT_FIELD_NAMES, 'page_context');
  const parsedOrganizationScope = organizationScope(raw.organization_scope);
  const parsedVisibility = visibility(raw.visibility);
  if (
    parsedVisibility === 'department' &&
    parsedOrganizationScope?.department_id == null
  ) {
    return fail('visibility', 'department visibility requires a department scope');
  }
  if (
    parsedVisibility === 'organization' &&
    parsedOrganizationScope?.organization_id == null
  ) {
    return fail(
      'visibility',
      'organization visibility requires an organization scope',
    );
  }

  return Object.freeze({
    surface_id: surfaceId(raw.surface_id),
    organization_scope: parsedOrganizationScope,
    work_object_refs: workObjectRefs(raw.work_object_refs),
    source_refs: sourceRefs(raw.source_refs),
    filters: filters(raw.filters),
    selected_metric: optionalIdentifier(raw.selected_metric, 'selected_metric'),
    allowed_capabilities: allowedCapabilities(raw.allowed_capabilities),
    freshness: freshness(raw.freshness),
    visibility: parsedVisibility,
  });
}

export function createGeneralPageContext(
  context: PageContextDeclaration,
): PageContextDeclaration {
  return parsePageContext({
    ...context,
    allowed_capabilities: [],
    source_refs: [],
    work_object_refs: [],
  });
}

File: app/ports/page_context.py
"""Fail-closed page-context registration and authorization contract."""

from __future__ import annotations

import json
import math
import re
from datetime import datetime
from typing import Annotated, Literal, NamedTuple, Protocol, TypeAlias
from weakref import WeakKeyDictionary

from pydantic import (
    AfterValidator,
    BaseModel,
    ConfigDict,
    Field,
    StrictBool,
    StrictFloat,
    StrictInt,
    StrictStr,
    StringConstraints,
    model_validator,
)

from app.ports.llm_provider import LLMMessage
from app.ports.trace import REDACTED_TRACE_VALUE, redact_trace_attributes

PAGE_CONTEXT_FIELD_NAMES = (
    "surface_id",
    "organization_scope",
    "work_object_refs",
    "source_refs",
    "filters",
    "selected_metric",
    "allowed_capabilities",
    "freshness",
    "visibility",
)
PAGE_CONTEXT_MAX_REFERENCES = 200
PAGE_CONTEXT_MAX_FILTERS = 32
PAGE_CONTEXT_MAX_CAPABILITIES = 64
PAGE_CONTEXT_TIMESTAMP_MAX_LENGTH = 64
PAGE_CONTEXT_TIMESTAMP_PATTERN = (
    r"^\d{4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[12]\d|3[01])"
    r"T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d{1,6})?Z$"
)
PAGE_CONTEXT_DATA_MESSAGE_PREFIX = "UNTRUSTED_PAGE_CONTEXT_DATA\n"

_SURFACE_ID_RE = re.compile(r"^[a-z0-9]+(?:[._-][a-z0-9]+)*$", re.ASCII)
_IDENTIFIER_RE = re.compile(r"^[A-Za-z0-9]+(?:[._:@/-][A-Za-z0-9]+)*$", re.ASCII)
_CAPABILITY_ID_RE = re.compile(r"^[a-z0-9]+(?:[._:-][a-z0-9]+)*$", re.ASCII)
_TIMESTAMP_RE = re.compile(PAGE_CONTEXT_TIMESTAMP_PATTERN, re.ASCII)
_DOM_MARKUP_RE = re.compile(
    r"<!doctype\s+html|<\s*/?\s*(?:html|body|form|input|script|table|div|main)\b",
    re.IGNORECASE,
)


class _PageContextModel(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        frozen=True,
        hide_input_in_errors=True,
    )


def _reject_credential_or_dom_text(value: str) -> str:
    if not value.isprintable():
        raise ValueError("page context text must contain only printable characters")
    redacted = redact_trace_attributes({"value": value})["value"]
    if redacted == REDACTED_TRACE_VALUE:
        raise ValueError("credential-like material is forbidden in page context")
    if _DOM_MARKUP_RE.search(value):
        raise ValueError("DOM markup is forbidden in page context")
    return value


def _validate_surface_id(value: str) -> str:
    if not _SURFACE_ID_RE.fullmatch(value):
        raise ValueError("surface_id must use lowercase slug form")
    return _reject_credential_or_dom_text(value)


def _validate_identifier(value: str) -> str:
    if not _IDENTIFIER_RE.fullmatch(value):
        raise ValueError("identifier contains unsupported characters")
    return _reject_credential_or_dom_text(value)


def _validate_capability_id(value: str) -> str:
    if not _CAPABILITY_ID_RE.fullmatch(value):
        raise ValueError("capability id must use lowercase dotted slug form")
    return _reject_credential_or_dom_text(value)


def _validate_timestamp(value: str) -> str:
    _reject_credential_or_dom_text(value)
    if not _TIMESTAMP_RE.fullmatch(value):
        raise ValueError("freshness observed_at must be a UTC RFC3339 timestamp")
    try:
        datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as exc:
        raise ValueError("freshness observed_at must be a valid timestamp") from exc
    return value


PageSurfaceId: TypeAlias = Annotated[
    StrictStr,
    StringConstraints(min_length=1, max_length=80),
    AfterValidator(_validate_surface_id),
]
PageContextIdentifier: TypeAlias = Annotated[
    StrictStr,
    StringConstraints(min_length=1, max_length=256),
    AfterValidator(_validate_identifier),
]
PageContextText: TypeAlias = Annotated[
    StrictStr,
    StringConstraints(min_length=1, max_length=500),
    AfterValidator(_reject_credential_or_dom_text),
]
PageContextReferenceText: TypeAlias = Annotated[
    StrictStr,
    StringConstraints(min_length=1, max_length=256),
    AfterValidator(_reject_credential_or_dom_text),
]
PageContextTimestamp: TypeAlias = Annotated[
    StrictStr,
    StringConstraints(min_length=1, max_length=PAGE_CONTEXT_TIMESTAMP_MAX_LENGTH),
    AfterValidator(_validate_timestamp),
]
PageCapabilityId: TypeAlias = Annotated[
    StrictStr,
    StringConstraints(min_length=1, max_length=120),
    AfterValidator(_validate_capability_id),
]
PageVisibility: TypeAlias = Literal["principal", "department", "organization"]
PageFilterOperator: TypeAlias = Literal[
    "equals",
    "not_equals",
    "contains",
    "greater_than",
    "less_than",
]
PageFilterValue: TypeAlias = (
    PageContextText | StrictInt | StrictFloat | StrictBool
)
PageFreshnessState: TypeAlias = Literal["unknown", "reported", "stale"]


class OrganizationScope(_PageContextModel):
    tenant_id: PageContextIdentifier
    organization_id: PageContextIdentifier | None
    department_id: PageContextIdentifier | None

    @model_validator(mode="after")
    def validate_department_parent(self) -> OrganizationScope:
        if self.department_id is not None and self.organization_id is None:
            raise ValueError("department_id requires organization_id")
        return self


class WorkObjectReference(_PageContextModel):
    work_object_id: PageContextIdentifier


class SourceReference(_PageContextModel):
    source_system: PageContextIdentifier
    source_ref: PageContextReferenceText


class PageFilter(_PageContextModel):
    """Untrusted page declaration; authority proves the exact control/value pair."""

    field: PageContextIdentifier
    operator: PageFilterOperator
    value: PageFilterValue
    source: Literal["visible_control"]

    @model_validator(mode="after")
    def validate_value(self) -> PageFilter:
        if isinstance(self.value, str):
            _reject_credential_or_dom_text(self.value)
        if isinstance(self.value, float) and not math.isfinite(self.value):
            raise ValueError("filter number must be finite")
        return self


class PageFreshness(_PageContextModel):
    state: PageFreshnessState
    observed_at: PageContextTimestamp | None

    @model_validator(mode="after")
    def validate_observation(self) -> PageFreshness:
        if self.state != "unknown" and self.observed_at is None:
            raise ValueError("reported or stale freshness requires observed_at")
        return self


class _PageContextData(_PageContextModel):
    surface_id: PageSurfaceId
    organization_scope: OrganizationScope | None
    work_object_refs: Annotated[
        tuple[WorkObjectReference, ...],
        Field(max_length=PAGE_CONTEXT_MAX_REFERENCES),
    ]
    source_refs: Annotated[
        tuple[SourceReference, ...],
        Field(max_length=PAGE_CONTEXT_MAX_REFERENCES),
    ]
    filters: Annotated[
        tuple[PageFilter, ...],
        Field(max_length=PAGE_CONTEXT_MAX_FILTERS),
    ]
    selected_metric: PageContextIdentifier | None
    allowed_capabilities: Annotated[
        tuple[PageCapabilityId, ...],
        Field(max_length=PAGE_CONTEXT_MAX_CAPABILITIES),
    ]
    freshness: PageFreshness
    visibility: PageVisibility

    @model_validator(mode="after")
    def validate_unique_collections(self) -> _PageContextData:
        work_object_ids = [item.work_object_id for item in self.work_object_refs]
        source_keys = [
            (item.source_system, item.source_ref) for item in self.source_refs
        ]
        if len(work_object_ids) != len(set(work_object_ids)):
            raise ValueError("work_object_refs must not contain duplicates")
        if len(source_keys) != len(set(source_keys)):
            raise ValueError("source_refs must not contain duplicates")
        if len(self.allowed_capabilities) != len(set(self.allowed_capabilities)):
            raise ValueError("allowed_capabilities must not contain duplicates")
        if self.visibility == "department" and (
            self.organization_scope is None
            or self.organization_scope.department_id is None
        ):
            raise ValueError("department visibility requires a department scope")
        if self.visibility == "organization" and (
            self.organization_scope is None
            or self.organization_scope.organization_id is None
        ):
            raise ValueError("organization visibility requires an organization scope")
        return self


class PageContextDeclaration(_PageContextData):
    """The only page-supplied shape; values remain untrusted declarations."""


class AuthorizedPageContext(_PageContextData):
    """Validated nine-field data; only a resolution ticket proves authorization."""


class PageContextAuthority(_PageContextModel):
    """Trusted facts built from Principal, Policy and Registry; never page input."""

    principal_id: PageContextIdentifier
    organization_scopes: frozenset[OrganizationScope]
    visibilities: frozenset[PageVisibility]
    registry_capabilities: frozenset[PageCapabilityId]
    policy_capabilities: frozenset[PageCapabilityId]
    visible_work_object_refs: frozenset[WorkObjectReference]
    visible_source_refs: frozenset[SourceReference]
    visible_filters: frozenset[PageFilter]
    visible_selected_metrics: frozenset[PageContextIdentifier]


class PageContextModelData(_PageContextModel):
    """Structured model input segment that can never become a system instruction."""

    role: Literal["user_data"] = "user_data"
    trust: Literal["untrusted_external"] = "untrusted_external"
    data: AuthorizedPageContext


class PageContextAuthorizationError(RuntimeError):
    """Explicit fail-closed authorization failure without rejected values."""

    def __init__(self, code: str) -> None:
        self.code = code
        super().__init__(code)


class _PageContextResolutionPayload(NamedTuple):
    principal_id: str
    authorized_context_json: str
    rejected_capabilities: tuple[str, ...]


class PageContextResolution:
    """Opaque authorization ticket issued only by ``authorize_page_context``."""

    __slots__ = ("_payload", "__weakref__")
    _payload: _PageContextResolutionPayload

    def __new__(cls, *args: object, **kwargs: object) -> PageContextResolution:
        del args, kwargs
        raise TypeError("PageContextResolution can only be issued by authorize_page_context")

    def __setattr__(self, name: str, value: object) -> None:
        del name, value
        raise TypeError("PageContextResolution is immutable")

    @property
    def principal_id(self) -> str:
        return self._payload.principal_id

    @property
    def authorized_context(self) -> AuthorizedPageContext:
        return AuthorizedPageContext.model_validate_json(
            self._payload.authorized_context_json
        )

    @property
    def rejected_capabilities(self) -> tuple[str, ...]:
        return self._payload.rejected_capabilities


class _AuthorizePageContext(Protocol):
    def __call__(
        self,
        declaration: PageContextDeclaration,
        authority: PageContextAuthority,
    ) -> PageContextResolution: ...


class _AsUntrustedModelData(Protocol):
    def __call__(
        self,
        resolution: PageContextResolution,
    ) -> PageContextModelData: ...


class _BuildPageContextMessages(Protocol):
    def __call__(
        self,
        *,
        system_instruction: str,
        resolution: PageContextResolution,
    ) -> tuple[LLMMessage, LLMMessage]: ...


def _build_page_context_authorization_contract() -> tuple[
    _AuthorizePageContext,
    _AsUntrustedModelData,
    _BuildPageContextMessages,
]:
    issued_resolutions: WeakKeyDictionary[
        PageContextResolution,
        _PageContextResolutionPayload,
    ] = WeakKeyDictionary()

    def require_issued_resolution(
        candidate: object,
    ) -> _PageContextResolutionPayload:
        if not isinstance(candidate, PageContextResolution):
            raise PageContextAuthorizationError("authorization_provenance_invalid")
        try:
            return issued_resolutions[candidate]
        except KeyError as exc:
            raise PageContextAuthorizationError(
                "authorization_provenance_invalid"
            ) from exc

    def authorize_page_context(
        declaration: PageContextDeclaration,
        authority: PageContextAuthority,
    ) -> PageContextResolution:
        """Authorize declarations and issue an identity-bound provenance ticket."""

        if (
            declaration.organization_scope is not None
            and declaration.organization_scope not in authority.organization_scopes
        ):
            raise PageContextAuthorizationError("organization_scope_not_authorized")
        if declaration.visibility not in authority.visibilities:
            raise PageContextAuthorizationError("visibility_not_authorized")
        if not set(declaration.work_object_refs).issubset(
            authority.visible_work_object_refs
        ):
            raise PageContextAuthorizationError("work_object_refs_not_visible")
        if not set(declaration.source_refs).issubset(authority.visible_source_refs):
            raise PageContextAuthorizationError("source_refs_not_visible")
        if not set(declaration.filters).issubset(authority.visible_filters):
            raise PageContextAuthorizationError("filters_not_visible")
        if (
            declaration.selected_metric is not None
            and declaration.selected_metric not in authority.visible_selected_metrics
        ):
            raise PageContextAuthorizationError("selected_metric_not_visible")

        backend_capabilities = (
            authority.registry_capabilities & authority.policy_capabilities
        )
        effective_capabilities = tuple(
            capability_id
            for capability_id in declaration.allowed_capabilities
            if capability_id in backend_capabilities
        )
        rejected_capabilities = tuple(
            capability_id
            for capability_id in declaration.allowed_capabilities
            if capability_id not in backend_capabilities
        )
        authorized_context = AuthorizedPageContext.model_validate(
            {
                **declaration.model_dump(mode="python"),
                "allowed_capabilities": effective_capabilities,
            }
        )
        payload = _PageContextResolutionPayload(
            principal_id=authority.principal_id,
            authorized_context_json=authorized_context.model_dump_json(),
            rejected_capabilities=rejected_capabilities,
        )
        resolution = object.__new__(PageContextResolution)
        object.__setattr__(resolution, "_payload", payload)
        issued_resolutions[resolution] = payload
        return resolution

    def as_untrusted_model_data(
        resolution: PageContextResolution,
    ) -> PageContextModelData:
        """Project only data from a factory-issued authorization ticket."""

        payload = require_issued_resolution(resolution)
        context = AuthorizedPageContext.model_validate_json(
            payload.authorized_context_json
        )
        return PageContextModelData(data=context)

    def build_page_context_messages(
        *,
        system_instruction: str,
        resolution: PageContextResolution,
    ) -> tuple[LLMMessage, LLMMessage]:
        """Assemble messages only from a factory-issued authorization ticket."""

        model_data = as_untrusted_model_data(resolution)
        serialized_data = json.dumps(
            model_data.model_dump(mode="json"),
            ensure_ascii=False,
            separators=(",", ":"),
        )
        return (
            LLMMessage(role="system", content=system_instruction),
            LLMMessage(
                role="user",
                content=f"{PAGE_CONTEXT_DATA_MESSAGE_PREFIX}{serialized_data}",
            ),
        )

    return (
        authorize_page_context,
        as_untrusted_model_data,
        build_page_context_messages,
    )


(
    authorize_page_context,
    as_untrusted_model_data,
    build_page_context_messages,
) = _build_page_context_authorization_contract()


class PageContextPort(Protocol):
    """Register/read context bound to principal_id; reads must reauthorize."""

    async def register(
        self,
        *,
        context: PageContextDeclaration,
        authority: PageContextAuthority,
    ) -> PageContextResolution: ...

    async def read(
        self,
        *,
        surface_id: PageSurfaceId,
        authority: PageContextAuthority,
    ) -> PageContextResolution | None: ...


__all__ = (
    "AuthorizedPageContext",
    "OrganizationScope",
    "PAGE_CONTEXT_FIELD_NAMES",
    "PAGE_CONTEXT_DATA_MESSAGE_PREFIX",
    "PAGE_CONTEXT_TIMESTAMP_MAX_LENGTH",
    "PAGE_CONTEXT_TIMESTAMP_PATTERN",
    "PageContextAuthorizationError",
    "PageContextAuthority",
    "PageContextDeclaration",
    "PageContextModelData",
    "PageContextPort",
    "PageContextResolution",
    "PageFilter",
    "PageFreshness",
    "SourceReference",
    "WorkObjectReference",
    "as_untrusted_model_data",
    "authorize_page_context",
    "build_page_context_messages",
)
